EFACs require EQUADs

In PINT, EFACs and EQUADs each operate on a subset of TOAs. But right now there is a restriction that each EFAC must have a corresponding EQUAD that operates on the same subset of TOAs. There is no reason for this restriction.

The following ``.par`` file produces an exception:
```
            PSRJ J1234+5678
            ELAT 0
            ELONG 0
            DM 10
            F0 1
            PEPOCH 58000
            EFAC mjd 57000 58000 2
```
The exception:
```
    def pair_EFAC_EQUAD(self):
        pairs = []
        for efac, efac_key in list(self.EFACs.items()):
            for equad, equad_key in list(self.EQUADs.items()):
                if efac_key == equad_key:
                    pairs.append((getattr(self, efac), getattr(self, equad)))
        if len(pairs) != len(list(self.EFACs.items())):
            # TODO may be define an parameter error would be helpful
>           raise ValueError(
                "Can not pair up EFACs and EQUADs, please "
                " check the EFAC/EQUAD keys and key values."
            )
E           ValueError: Can not pair up EFACs and EQUADs, please  check the EFAC/EQUAD keys and key values.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,7 @@ and this project, at least loosely, adheres to [Semantic Versioning](https://sem
 - Fixed bug that prevented reading Parkes-format TOAs
 - Fixed bug in solar wind model that prevented fitting
 - Fix pintempo script so it will respect JUMPs in the TOA file.
+- Uncertainties are no longer set to zero if some TOAs lack EFACs. (PR #890)
 ### Added
 - Added metadata to observatory definition, to keep track of the data origin
 - Added other bipm???? files from TEMPO2
@@ -39,6 +40,7 @@ modified after a fit and are displayed with a model's .par file output.
 - Changed requirements to astropy>=4.0
 - get_model can now read from file-like, including StringIO, objects (handy for testing) (PR #871)
 - WidebandDMResiduals now support access to their parts through .toa and .dm attributes (PR #861)
+- EFACs and EQUADs can be set independently from each other now. (PR #890)
 
 ## [0.7.0] - 2020-05-27
 ### Changed

--- a/src/pint/models/noise_model.py
+++ b/src/pint/models/noise_model.py
@@ -97,7 +97,7 @@ class ScaleToaError(NoiseComponent):
                 continue
         # convert all the TNEQ to EQUAD
 
-        for tneq in list(self.TNEQs.keys()):
+        for tneq in self.TNEQs:
             tneq_par = getattr(self, tneq)
             if tneq_par.key is None:
                 continue
@@ -142,33 +142,17 @@ class ScaleToaError(NoiseComponent):
             if [x for x in l if l.count(x) > 1] != []:
                 raise ValueError("'%s' have duplicated keys and key values." % el)
 
-    # pairing up EFAC and EQUAD
-    def pair_EFAC_EQUAD(self):
-        pairs = []
-        for efac, efac_key in list(self.EFACs.items()):
-            for equad, equad_key in list(self.EQUADs.items()):
-                if efac_key == equad_key:
-                    pairs.append((getattr(self, efac), getattr(self, equad)))
-        if len(pairs) != len(list(self.EFACs.items())):
-            # TODO may be define an parameter error would be helpful
-            raise ValueError(
-                "Can not pair up EFACs and EQUADs, please "
-                " check the EFAC/EQUAD keys and key values."
-            )
-        return pairs
-
     def scale_toa_sigma(self, toas):
-        tbl = toas.table
-        sigma_old = tbl["error"].quantity
-        sigma_scaled = np.zeros_like(sigma_old)
-        EF_EQ_pairs = self.pair_EFAC_EQUAD()
-        for pir in EF_EQ_pairs:
-            efac = pir[0]
-            equad = pir[1]
-            mask = efac.select_toa_mask(toas)
-            sigma_scaled[mask] = efac.quantity * np.sqrt(
-                sigma_old[mask] ** 2 + (equad.quantity) ** 2
-            )
+        sigma_scaled = toas.table["error"].quantity.copy()
+        for equad_name in self.EQUADs:
+            equad = getattr(self, equad_name)
+            if equad.quantity is None:
+                continue
+            mask = equad.select_toa_mask(toas)
+            sigma_scaled[mask] = np.hypot(sigma_scaled[mask], equad.quantity)
+        for efac_name in self.EFACs:
+            efac = getattr(self, efac_name)
+            sigma_scaled[efac.select_toa_mask(toas)] *= efac.quantity
         return sigma_scaled
 
     def sigma_scaled_cov_matrix(self, toas):

--- a/src/pint/models/parameter.py
+++ b/src/pint/models/parameter.py
@@ -196,8 +196,7 @@ class Parameter(object):
     # Setup quantity property
     @property
     def quantity(self):
-        """Return the internal stored parameter value and units.
-        """
+        """Return the internal stored parameter value and units."""
         return self._quantity
 
     @quantity.setter
@@ -266,8 +265,7 @@ class Parameter(object):
 
     @property
     def uncertainty(self):
-        """Return the internal stored parameter uncertainty value and units.
-        """
+        """Return the internal stored parameter uncertainty value and units."""
         return self._uncertainty
 
     @uncertainty.setter
@@ -419,8 +417,7 @@ class Parameter(object):
         return True
 
     def name_matches(self, name):
-        """Whether or not the parameter name matches the provided name
-        """
+        """Whether or not the parameter name matches the provided name"""
         return (name == self.name.upper()) or (
             name in map(lambda x: x.upper(), self.aliases)
         )
@@ -714,8 +711,7 @@ class boolParameter(Parameter):
         self.paramType = "boolParameter"
 
     def set_quantity_bool(self, val):
-        """ This function is to get boolean value for boolParameter class
-        """
+        """This function is to get boolean value for boolParameter class"""
         # First try strings
         try:
             if val.upper() in ["Y", "YES", "T", "TRUE", "1"]:
@@ -837,10 +833,10 @@ class MJDParameter(Parameter):
 
     def set_quantity_mjd(self, val):
         """Value setter for MJD parameter,
-           Accepted format:
-           Astropy time object
-           mjd float
-           mjd string (in pulsar_mjd format)
+        Accepted format:
+        Astropy time object
+        mjd float
+        mjd string (in pulsar_mjd format)
         """
         if isinstance(val, numbers.Number):
             val = np.longdouble(val)
@@ -952,7 +948,7 @@ class AngleParameter(Parameter):
         )
 
     def set_quantity_angle(self, val):
-        """ This function is to set value to angle parameters.
+        """This function is to set value to angle parameters.
 
         Accepted format:
         1. Astropy angle object
@@ -973,8 +969,7 @@ class AngleParameter(Parameter):
         return result
 
     def set_uncertainty_angle(self, val):
-        """This function is to set the uncertainty for an angle parameter.
-        """
+        """This function is to set the uncertainty for an angle parameter."""
         if isinstance(val, numbers.Number):
             result = Angle(val * self.unit_identifier[self._str_unit.lower()][2])
         elif isinstance(val, str):
@@ -990,16 +985,14 @@ class AngleParameter(Parameter):
         return result
 
     def print_quantity_angle(self, quan):
-        """This is a function to print out the angle parameter.
-        """
+        """This is a function to print out the angle parameter."""
         if ":" in self._str_unit:
             return quan.to_string(sep=":", precision=8)
         else:
             return quan.to_string(decimal=True, precision=15)
 
     def print_uncertainty(self, unc):
-        """This is a function for printing out the uncertainty
-        """
+        """This is a function for printing out the uncertainty"""
         if ":" in self._str_unit:
             angle_arcsec = unc.to(u.arcsec)
             if self.units == u.hourangle:
@@ -1285,21 +1278,32 @@ class prefixParameter(object):
 
 
 class maskParameter(floatParameter):
-    """Mask parameters are ones that apply to a subset of TOAs.
+    """Parameter that applies to a subset of TOAs.
 
-    This is a Parameter type for mask parameters which is to select a
-    certain subset of TOAs and apply changes on the subset of TOAs, for example
-    JUMP. This type of parameter does not require index input. But eventrully
-    an index part will be added, for the purpose of parsing the right value
-    from the parfile. For example::
+    A maskParameter applies to a subset of the TOAs, for example JUMP specifies
+    that their arrival times should be adjusted by the value associated with
+    this JUMP. The criterion is based on either one of the standard fields
+    (telescope, frequency, et cetera) or a flag; and the selection can be on an
+    exact match or on a range.
 
-        >>> p = maskParameter(name='JUMP', index=2)
+    Upon creation of a maskParameter, an index part will be added, so that the
+    parameters can be distinguished within the
+    :class:`pint.models.timing_model.TimingModel` object. For example::
+
+        >>> p = maskParameter(name='JUMP', index=2, key="-fe", key_value="G430")
         >>> p.name
         'JUMP2'
 
+    The selection criterion can be one of the parameters ``mjd``, ``freq``,
+    ``name``, ``tel`` representing the required columns of a ``.tim`` file, or
+    the name of a flag, starting with ``-``. If the selection criterion is
+    based on ``mjd`` or ``freq`` it is expected to be accompanied by a pair of
+    values that define a range; other criteria are expected to be accompanied
+    by a string that is matched exactly.
+
     Parameters
     ----------
-    name : str, optional
+    name : str
         The name of the parameter.
     index : int, optional
         The index number for the prefixed parameter.
@@ -1326,15 +1330,13 @@ class maskParameter(floatParameter):
         Whether derivatives with respect to this parameter make sense.
     aliases : list, optional
         List of aliases for parameter name.
-
-    TODO: Is mask parameter provide some other type of parameters other then
-    floatParameter?
-
     """
+
+    # TODO: Is mask parameter provide some other type of parameters other then floatParameter?
 
     def __init__(
         self,
-        name=None,
+        name,
         index=1,
         key=None,
         key_value=None,
@@ -1347,13 +1349,13 @@ class maskParameter(floatParameter):
         continuous=False,
         aliases=[],
     ):
-        self.is_mask = True
         self.key_identifier = {
             "mjd": (lambda x: time.Time(x, format="mjd").mjd, 2),
             "freq": (float, 2),
             "name": (str, 1),
             "tel": (str, 1),
         }
+        self.is_mask = True
         if key_value is None:
             key_value = []
         elif not isinstance(key_value, list):
@@ -1446,7 +1448,6 @@ class maskParameter(floatParameter):
             NAME key key_value1 key_value2 parameter_value uncertainty
 
         where NAME is the name for this class as reported by ``self.name_matches``.
-
         """
         k = line.split()
         if not k:
@@ -1528,8 +1529,7 @@ class maskParameter(floatParameter):
         return line + "\n"
 
     def new_param(self, index, copy_all=False):
-        """Create a new but same style mask parameter
-        """
+        """Create a new but same style mask parameter"""
         if not copy_all:
             new_mask_param = maskParameter(
                 name=self.origin_name,
@@ -1556,13 +1556,16 @@ class maskParameter(floatParameter):
         return new_mask_param
 
     def select_toa_mask(self, toas):
-        """Select the toas.
+        """Select the toas that match the mask.
+
         Parameter
         ---------
-        toas : TOAs class
-        Return
-        ------
-        A array of returned index.
+        toas: :class:`pint.toas.TOAs`
+
+        Returns
+        -------
+        array
+            An array of TOA indices selected by the mask.
         """
         column_match = {"mjd": "mjd_float", "freq": "freq", "tel": "obs"}
         if len(self.key_value) == 1:
@@ -1573,6 +1576,8 @@ class maskParameter(floatParameter):
             if not hasattr(self, "toa_selector"):
                 self.toa_selector = TOASelect(is_range=True, use_hash=True)
             condition = {self.name: tuple(self.key_value)}
+        elif len(self.key_value) == 0:
+            return np.array([], dtype=int)
         else:
             raise ValueError(
                 "Parameter %s has more key values than "

--- a/tests/test_residuals.py
+++ b/tests/test_residuals.py
@@ -2,6 +2,7 @@
 """
 
 import os
+from io import StringIO
 
 import astropy.units as u
 import numpy as np
@@ -11,7 +12,7 @@ from pinttestdata import datadir
 from pint.models import get_model
 from pint.models.dispersion_model import Dispersion
 from pint.residuals import CombinedResiduals, Residuals, WidebandTOAResiduals
-from pint.toa import get_TOAs
+from pint.toa import get_TOAs, make_fake_toas
 from pint.utils import weighted_mean
 
 os.chdir(datadir)
@@ -20,6 +21,7 @@ os.chdir(datadir)
 class TestResidualBuilding:
     def setup(self):
         self.model = get_model("J1614-2230_NANOGrav_12yv3.wb.gls.par")
+        # self.toa = make_fake_toas(57000,59000,20,model=self.model)
         self.toa = get_TOAs("J1614-2230_NANOGrav_12yv3.wb.tim")
 
     def test_build_phase_residual(self):
@@ -88,3 +90,26 @@ class TestResidualBuilding:
         # Make sure the model object are shared by all individual residual class
         assert wb_res.model is wb_res.toa.model
         assert wb_res.model is wb_res.dm.model
+
+
+def test_residuals_scaled_uncertainties():
+    model = get_model(
+        StringIO(
+            """
+            PSRJ J1234+5678
+            ELAT 0
+            ELONG 0
+            DM 10
+            F0 1
+            PEPOCH 58000
+            EFAC mjd 57000 58000 2
+            """
+        )
+    )
+    toas = make_fake_toas(57000, 59000, 20, model=model, error=1 * u.us)
+    r = Residuals(toas, model)
+    e = r.get_data_error(scaled=True)
+    assert np.all(e != 0)
+    assert 0 < np.sum(e > 1.5 * u.us) < len(toas)
+    with pytest.raises(ValueError):
+        model.as_parfile().index("EQUAD")

